### PR TITLE
OpenAI Conversation: Lookup

### LINKED ISSUE #279: OpenAI: Response API Conversation Memory
**Is your feature request related to a problem? Please describe.**
We have recently integrated OpenAI response API in AI Platform but to retain conversation memory there is a slight twist as opposed to threads API
According to the docs here:
https://platform.openai.com/docs/guides/conversation-state?api-mode=responses#openai-apis-for-conversation-state

If we extend the code to send a third message, we’re supposed to pass the **response_id** from the **second_response**. But this **response_id** is different from the response_id from **first_response** — it has the same prefix and suffix, but the middle part changes.

This seems quite different from the Threads API, where one can keep the same thread_id for the whole conversation, even though each message has its own message_id.

**Describe the solution you'd like**
While this is raised to OpenAI team, we'll need a temporary solution meanwhile in AI Platform to retain conversation memory through single ID as we are planning to migrate Glific

### LINKED ISSUE #304: OpenAI Conversation: Conversation Memory Lookup
**Is your feature request related to a problem? Please describe.**
We’re seeing an issue with the Response API. The conversation_id seems to change with every message.

According to the docs here:
https://platform.openai.com/docs/guides/conversation-state?api-mode=responses#openai-apis-for-conversation-state

If we extend the code to send a third message, we’re supposed to pass the response_id from the second_response. But this response_id is different from the response_id from first_response — it has the same prefix and suffix, but the middle part changes.

This seems quite different from the Threads API, where one can keep the same thread_id for the whole conversation, even though each message has its own message_id.

**Describe the solution you'd like**
Thus to mitigate this issue we have to store conversation history at our end as we have added CRUD we need to add a logic to lookup previous responses to retain conversation memory

### PR DESCRIPTION
**Pull request by AkhileshNegi**
## Summary

Target issue is #304
Explain the **motivation** for making this change. What existing problem does the pull request solve?

## Checklist

Before submitting a pull request, please ensure that you mark these task.

- [x] Ran `fastapi run --reload app/main.py` or `docker compose up` in the repository root and test.
- [x] If you've fixed a bug or added code that is tested and has test cases.

## Notes

Please add here if any other information is required for the reviewer.

## Summary by CodeRabbit

* **New Features**
  * Enhanced conversation tracking by linking responses with ancestor conversations for improved conversation history and context.
  * Both asynchronous and synchronous response endpoints now persist conversation ancestry and metadata for better traceability.

* **Bug Fixes**
  * Improved error message handling for more robust and informative feedback when issues occur.

* **Tests**
  * Expanded and improved test coverage for conversation ancestry logic and response endpoints, including handling of edge cases and error scenarios.

### CODE REVIEW

File: backend/app/api/routes/openai_conversation.py [hunk 1]
```diff
@@ -15,6 +15,8 @@
 )
 from app.models import (
     UserProjectOrg,
+    OpenAIConversationCreate,
+    OpenAIConversation,
```
Comment (avirajsingh7) [inline backend/app/api/routes/openai_conversation.py]: remove unused imports

File: backend/app/api/routes/responses.py [hunk 2]
```diff
@@ -117,9 +136,20 @@ def process_response(
...
     try:
+        # Get the latest conversation by ancestor ID to use as previous_response_id
+        previous_response_id = request.response_id
+        if request.response_id:
+            latest_conversation = get_conversation_by_ancestor_id(
+                session=session,
+                ancestor_response_id=request.response_id,
+                project_id=project_id,
+            )
+            if latest_conversation:
```
Comment (avirajsingh7) [inline backend/app/api/routes/responses.py:148]: if request.response_id is provided, that means there should be previous conversation with respect to that response_id,
If conversation not exist its kind of invalid response id, instead of generating result which will use incorrect previous_response_id, throw exception

What do you think? (openAI may also handle this, up to you to decide best design here)

File: backend/app/api/routes/responses.py [hunk 3]
```diff
@@ -166,6 +196,40 @@ def process_response(
...
+            create_conversation(
+                session=session,
+                conversation=conversation_data,
+                project_id=project_id,
+                organization_id=organization_id,
+            )
+            logger.info(
+                f"Created conversation record for response_id={response.id}, assistant_id={mask_string(request.assistant_id)}, project_id={project_id}"
+            )
+        except Exception as e:
```
Comment (avirajsingh7) [inline backend/app/api/routes/responses.py]: Do we need try catch here?

File: backend/app/crud/openai_conversation.py [hunk 4]
```diff
@@ -57,6 +58,47 @@ def get_conversation_by_ancestor_id(
     return result
 
 
+def set_ancestor_response_id(
```
Comment (avirajsingh7) [inline backend/app/crud/openai_conversation.py]: instead of set this should be `get_ancestor_id_from_response_id`

**Review by avirajsingh7 — CHANGES_REQUESTED** [review]

File: backend/app/api/routes/responses.py [hunk 5]
```diff
@@ -166,6 +196,40 @@ def process_response(
             },
         )
 
+        # Set ancestor_response_id using CRUD function
+        ancestor_response_id = set_ancestor_response_id(
```
Comment (avirajsingh7) [inline backend/app/api/routes/responses.py]: Do we need this,
we can take ancestor_Id from `latest_conversation` and if `latest_conversation` none than use response id as ancestor_Id

**Review by avirajsingh7 — CHANGES_REQUESTED** [review]

File: backend/app/crud/openai_conversation.py (hunk 4 above)
Comment (AkhileshNegi) [reply backend/app/crud/openai_conversation.py]: updated named

File: backend/app/api/routes/openai_conversation.py (hunk 1 above)
Comment (AkhileshNegi) [reply backend/app/api/routes/openai_conversation.py]: removed

File: backend/app/api/routes/responses.py (hunk 2 above)
Comment (AkhileshNegi) [reply backend/app/api/routes/responses.py:148]: that might not be the case as at flow level people can send all kind of things. And what we do is being the intermediatory between them and OpenAI. OpenAI does throw error which we send in callback
1. when it is random string
image

2. random string that starts with resp_
image

But do let me know your thoughts

File: backend/app/api/routes/responses.py (hunk 3 above)
Comment (AkhileshNegi) [reply backend/app/api/routes/responses.py]: nope, so removed it now

File: backend/app/api/routes/responses.py (hunk 5 above)
Comment (AkhileshNegi) [reply backend/app/api/routes/responses.py]: that's what that function does

Comment (avirajsingh7) [reply backend/app/api/routes/responses.py]: @AkhileshNegi  instead of calling a function to do that, you already fetched `latest_conversation` you can use that,
reducing number of db call

File: backend/app/api/routes/responses.py [hunk 6]
```diff
@@ -165,6 +196,35 @@ def process_response(
                 "error": None,
             },
         )
+        # Set ancestor_response_id using CRUD function
+        ancestor_response_id = (
+            latest_conversation.ancestor_response_id
+            if latest_conversation
+            else get_ancestor_id_from_response(
```
Comment (avirajsingh7) [inline backend/app/api/routes/responses.py:203]: instead of get_ancestor_id_from_response, can we just use response.id here?

File: backend/app/api/routes/responses.py [hunk 7]
```diff
@@ -117,9 +136,21 @@ def process_response(
     )
 
     try:
+        # Get the latest conversation by ancestor ID to use as previous_response_id
+        previous_response_id = request.response_id
```
Comment (avirajsingh7) [inline backend/app/api/routes/responses.py]: change `previous_response_id` variable name to `ancestor_id`, kind of mis leading and confusing as previous_response_id

**Review by avirajsingh7 — CHANGES_REQUESTED** [review]

File: backend/app/api/routes/responses.py (hunk 6 above)
Comment (AkhileshNegi) [reply backend/app/api/routes/responses.py:203]: No, that will break down the line.. like even for 2nd and 3rd response the ancestor id should be 1st while previous_response_id for 2nd will be 1st and previous_response_id for 3rd will be 2nd and

File: backend/app/api/routes/responses.py (hunk 7 above)
Comment (AkhileshNegi) [reply backend/app/api/routes/responses.py]: updated

**Review by avirajsingh7 — APPROVED** [review]
